pkg/debug/spec.sh

```diff
@@ -324,11 +267,129 @@ __EOT__
+    # passthrough of the device into an edge application
+    # fortunately this is not a problem as we currently disallow the passthrough of
+    # usb networking devices in order to not confuse pillar daemon
+    # no kernel module -> pillar does not get confused -> passthrough of the device is okay
+    for i in /sys/class/net/*
+    do
+        i=$(basename "$i")
+        local devicepath
+        devicepath=$(realpath "/sys/class/net/$i/../../")
+        local isUSB
+        isUSB=$(echo "$devicepath" | grep -Eo '/usb[0-9]/' || true)
+
+        if [ "$isUSB" = "" ];
+        then
+            continue
+        fi
+        local netdevpaths="$netdevpaths $devicepath"
+    done
+
+    for i in $(find /sys/devices/ -name uevent | grep -E '/usb[0-9]/' | grep -E '/[0-9]-[0-9](\.[0-9]+)?/uevent')
+    do
+        local ignore_dev=0
+        local devicepath
+        devicepath=$(dirname "$i")
+        local busAndPort
+        busAndPort=$(echo "$i" | grep -Eo '/[0-9]-[0-9](\.[0-9]+)?/uevent' | grep -Eo '[0-9]-[0-9](\.[0-9]+)?')
+        local assigngrp
+        assigngrp="USB${busAndPort}"
+
+        local bus
+        bus=$(echo "${busAndPort}" | cut -d - -f1)
+        local port
+        port=$(echo "${busAndPort}" | cut -d - -f2)
+        local usbaddr="$bus:$port"
+
+        local ifname
+        ifname="$(ls -1 "${devicepath}/*/net" 2>/dev/null || true)"
+        local cost=0
+        local isModem
+        isModem=$(grep -Eo 'cdc_mbim|qmi_wwan' "${devicepath}"/*/uevent 2>/dev/null)
+        if [ "${isModem}" != "" ]
+        then
+            cost=10
+            assigngrp="modem${busAndPort}"
+            assigngrp=$(get_assignmentgroup "${ifname}" "${pciaddr}")
+        fi
+
+        local pciaddr
+        pciaddr=$(echo "$i" | grep -Eo '/pci[^/]+/[^/]+/usb' | cut -d / -f3)
+        local parentassigngrp
+        parentassigngrp="group$(pci_iommu_group "${pciaddr}")"
+        if [ "$parentassigngrp" = "group" ]
+        then
+            parentassigngrp=""
+        fi
+        local product
+        product=$(cat "$devicepath/product" 2> /dev/null || true)
+        local label
+        label=$(echo "$product" | tr -cd '[:alnum:]._-')
+        if [ "$label" = "" ]
+        then
+            label="USB${busAndPort}"
+        fi
+
+        type=$(grep -Eo '^TYPE=[0-9]+/' "$i"| grep -Eo '[0-9]+')
+        # ignore USB hubs
+        if [ "$type" = "9" ];
+        then
+            ignore_dev=1
+        fi
+
+        # ignore network cards but allow usb modems
+        if [ "${isModem}" = "" ]
+        then
+            for netdevpath in $netdevpaths
+            do
+                # check if devicepath starts with netdevpath
+                case $netdevpath in "$devicepath"*)
+                    ignore_dev=1
+                esac
+            done
+        fi
+
+        if [ "$ignore_dev" = "1" ]
+        then
+            continue
+        fi
+
+        cat <<__EOT__
+    {
+      "ztype": "IO_TYPE_UNSPECIFIED",
```

Comment:
Shouldn't this be IO_TYPE_USB (for non-modem device) or IO_TYPE_WWAN (for cellular modem)?

Reply:
`IO_TYPE_USB` will be added in a later PR once parent assign groups are supported.

Regarding `IO_TYPE_WWAN` I am unsure; why would we care what kind of usb device we passthrough into the edge application?

Besides `IO_TYPE_WWAN` I also found `IO_TYPE_LTE`.

Reply:
Note that listing modems by spec.sh is not only for passthrough purposes. Controller needs to now about all network adapters (IO_TYPE_ETH + IO_TYPE_WWAN), so that they can have network config attached and used for device connectivity purposes.
And ignore IO_TYPE_LTE, this was apparently superseded by IO_TYPE_WWAN (LTE is only one generation of cellular connectivity, but 3G or 5G can be used as well, so LTE is not a generic enough label for modems).